File: apps/api/app/schemas/schemes.py
from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import HttpUrl

from app.models.enums import ReviewStatus
from app.schemas.common import APIModel


class LocalizedTextOut(APIModel):
    en: str
    te: str


class SchemeSourceOut(APIModel):
    """Provenance for one official scheme claim."""

    source_record_id: UUID
    source_name: str
    official_source_url: HttpUrl
    public_source_url: HttpUrl | None = None
    retrieval_date: date
    review_status: ReviewStatus


class SchemeClaimOut(APIModel):
    """An official single-value scheme claim with its source."""

    classification: Literal["official"]
    value: LocalizedTextOut
    source: SchemeSourceOut


class SchemeListClaimOut(APIModel):
    """An official multi-value scheme claim with its source."""

    classification: Literal["official"]
    value: list[LocalizedTextOut]
    source: SchemeSourceOut


class SchemeRecordOut(APIModel):
    """One reviewed Andhra Pradesh scheme with per-claim provenance."""

    slug: str
    name: SchemeClaimOut
    description: SchemeClaimOut
    category: SchemeClaimOut
    department: SchemeClaimOut | None = None
    districts: SchemeListClaimOut | None = None
    eligibility: SchemeListClaimOut | None = None


class SchemeCatalogOut(APIModel):
    data: list[SchemeRecordOut]
    status: Literal["prepared-empty", "reviewed"]
    telugu_reviewed: bool

File: apps/web/src/lib/schemes.ts
import type { Locale } from "./catalog-types";

export interface LocalizedText {
  en: string;
  te: string;
}

export interface SchemeSourceRecord {
  source_record_id: string;
  source_name: string;
  official_source_url: string;
  retrieval_date: string;
  review_status: "reviewed";
}

export interface OfficialSchemeClaim<T> {
  classification: "official";
  value: T;
  source: SchemeSourceRecord;
}

export interface SchemeRecord {
  slug: string;
  name: OfficialSchemeClaim<LocalizedText>;
  description: OfficialSchemeClaim<LocalizedText>;
  department: OfficialSchemeClaim<LocalizedText>;
  districts: OfficialSchemeClaim<LocalizedText[]>;
  category: OfficialSchemeClaim<LocalizedText>;
  eligibility: OfficialSchemeClaim<LocalizedText[]> | null;
}

export interface SchemeFilters {
  department: string;
  district: string;
  category: string;
  eligibility: "all" | "published" | "unavailable";
}

export interface SchemeCatalogResponse {
  data: SchemeRecord[];
  status: "prepared-empty" | "reviewed";
}

// Production remains empty until records complete source and bilingual review.
export const preparedSchemes: readonly SchemeRecord[] = [];

export function localized(value: LocalizedText, locale: Locale): string {
  return locale === "te" ? value.te : value.en;
}

export function filterSchemes(
  schemes: SchemeRecord[],
  filters: SchemeFilters,
): SchemeRecord[] {
  return schemes.filter((scheme) => {
    const hasEligibility = scheme.eligibility !== null;
    return (
      (!filters.department ||
        scheme.department.value.en === filters.department) &&
      (!filters.district ||
        scheme.districts.value.some(
          (district) => district.en === filters.district,
        )) &&
      (!filters.category || scheme.category.value.en === filters.category) &&
      (filters.eligibility === "all" ||
        (filters.eligibility === "published" && hasEligibility) ||
        (filters.eligibility === "unavailable" && !hasEligibility))
    );
  });
}

export function preparedSchemeBySlug(slug: string): SchemeRecord | null {
  return preparedSchemes.find((scheme) => scheme.slug === slug) ?? null;
}

export async function getSchemes(
  signal?: AbortSignal,
): Promise<SchemeCatalogResponse> {
  const response = await fetch("/api/schemes", {
    headers: { Accept: "application/json" },
    signal,
  });
  if (!response.ok) {
    throw new Error(`Scheme request failed with status ${response.status}`);
  }
  return (await response.json()) as SchemeCatalogResponse;
}
